agent/checks/alias.go

```diff
@@ -0,0 +1,198 @@
+package checks
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/types"
+)
+
+// Constants related to alias check backoff.
+const (
+	checkAliasBackoffMin     = 3               // 3 attempts before backing off
+	checkAliasBackoffMaxWait = 1 * time.Minute // maximum backoff wait time
+)
+
+// CheckAlias is a check type that aliases the health of another service
+// instance or node. If the service aliased has any critical health checks, then
+// this check is critical. If the service has no critical but warnings,
+// then this check is warning, and if a service has only passing checks, then
+// this check is passing.
+type CheckAlias struct {
+	Node      string // Node name of the service. If empty, assumed to be this node.
+	ServiceID string // ID (not name) of the service to alias
+
+	CheckID types.CheckID               // ID of this check
+	RPC     RPC                         // Used to query remote server if necessary
+	RPCReq  structs.NodeSpecificRequest // Base request
+	Notify  AliasNotifier               // For updating the check state
+
+	stop     bool
+	stopCh   chan struct{}
+	stopLock sync.Mutex
+}
+
+// AliasNotifier is a CheckNotifier specifically for the Alias check.
+// This requires additional methods that are satisfied by the agent
+// local state.
+type AliasNotifier interface {
+	CheckNotifier
+
+	AddAliasCheck(types.CheckID, string, chan<- struct{}) error
+	RemoveAliasCheck(types.CheckID, string)
+	Checks() map[types.CheckID]*structs.HealthCheck
+}
+
+// Start is used to start the check, runs until Stop() func (c *CheckAlias) Start() {
+func (c *CheckAlias) Start() {
+	c.stopLock.Lock()
```

Comment:
Agreed. I just copy/pasted the other checks. When we have some more breathing room I was going to propose unifying both this as well as a check interface so that we can have less repetition elsewhere (in the agent and config parsers).

Reply:
I just whipped up a new type to mimic the Pythone threading.Event type (https://docs.python.org/2.7/library/threading.html#event-objects) with an extra function to get at the actual channel to wait on if asked. 

https://gist.github.com/mkeeler/cb88cc762ca36733db0798ca80f1e73e

We could use that or do something else more sophisticated.

Reply:
Nice, so this is a bigger discussion for outside of this thread, but we've generally avoided abstracting "shutdown" channels because the shutdown behavior you want is generally different by type of thing. Abstracting shutdown for _checks_ though (smaller scope with equivalent behavior) is reasonable to me, but we should talk about it more later when I'm back.